src/_P092_DLbus.ino

```diff
@@ -644,61 +672,66 @@ boolean Plugin_092_Receiving(void) {
   uint8_t rawval, val;
   int i;
   noInterrupts ();                              // make sure we don't get interrupted before we are ready
-  pulse_count = 0;
-  pulse_number = (((DataSettings.Datauint8_ts + AdditionalRecuint8_ts) * (8 + 1 + 1) + 16) * 4);
-  MinPulseWidth = DataSettings.MinPulseWidth;
-  MaxPulseWidth = DataSettings.MaxPulseWidth;
-  MinDoublePulseWidth = DataSettings.MinDoublePulseWidth;
-  MaxDoublePulseWidth = DataSettings.MaxDoublePulseWidth;
-  addLog(LOG_LEVEL_INFO, F("P092_receiving ..."));
+  DLbus_pulse_count = 0;
+  DLbus_pulse_number = (((P092_DataSettings.DataBytes + DLbus_AdditionalRecBytes) * (8 + 1 + 1) + 16) * 3);
+  DLbus_MinPulseWidth = P092_DataSettings.DLbus_MinPulseWidth;
+  DLbus_MaxPulseWidth = P092_DataSettings.DLbus_MaxPulseWidth;
+  DLbus_MinDoublePulseWidth = P092_DataSettings.DLbus_MinDoublePulseWidth;
+  DLbus_MaxDoublePulseWidth = P092_DataSettings.DLbus_MaxDoublePulseWidth;
   P092_receiving = true;
   interrupts ();                                // interrupts allowed now, next instruction WILL be executed
   uint32_t start=millis();
-  while ((millis()-start)<100) {
-    // wait 100 ms
+  addLog(LOG_LEVEL_INFO, F("P092_receiving ..."));
+  while ((timePassedSince(start)<100) && (DLbus_pulse_count == 0)) {
+    // wait for first pulse received (timeout 100ms)
     yield();
   }
-  if (pulse_count == 0) {
+  if (DLbus_pulse_count == 0) {
     // nothing received
     P092_receiving = false;
     addLog(LOG_LEVEL_ERROR, F("Error: Nothing received! No DL bus connected!"));
     return false;
   }
   while (P092_receiving) {
     // wait for end of P092_receiving
+    // loop time depends on the device (bytes in protocol and clock frequency)
+    // ESR21:     (((31 + 2) * (8 + 1 + 1) + 16) * 3) = 1038 bits @ 488Hz = 2,13s
+    // UVR31:     ((( 8 + 2) * (8 + 1 + 1) + 16) * 3) =  348 bits @  50Hz = 6,96s
+    // UVR1611:   (((64 + 2) * (8 + 1 + 1) + 16) * 3) = 2028 bits @ 488Hz = 4,16s
+    // UVR 61-3:  (((62 + 2) * (8 + 1 + 1) + 16) * 3) = 2028 bits @ 488Hz = 4,03s
```

Comment:
Do you really need to actively wait here?
waiting for several seconds is a big no-no as you are blocking all other functionality of the node.
Please have a good look at what you are waiting for here.
The received bits will be processed in an interrupt-driven ISR call, so why wait here?
This can typically be handled in the 50/sec or 10/sec call and if a reading is finished you simply re-schedule the task to perform a PLUGIN_READ to collect the data.

Please have a look at the code of BME280.

Reply:
I know, that this long blocking of all other functionalities is very bad, but I'm afraid that I loose some more bit change information while running other stuff as I already encounter.
**The if-condition on line 711 is already handling (that means guessing!) lost bit changes.**
So far the code runs fine on my WEMOS D1R1 mini for more than 1 year.
I just use two devices (DL bus and Sysinfo). The data of both devices are send via http to my FHEM home automation.
Nevertheless I'll try another code on my device (ESR21) coming weekend.
By the way I got some other questions:

- How big is the IRAM on the ESP82xx devices?

- Is there a priority for the interrupts?

- Can an interrupt function be interrupted (e.g. `attachInterrupt(digitalPinToInterrupt(P092_DLB_Pin), P092_Pin_changed, CHANGE)`)
by another interrupt?

Reply:
> By the way I got some other questions:
> 
> * How big is the IRAM on the ESP82xx devices?
> * Is there a priority for the interrupts?
> * Can an interrupt function be interrupted (e.g. `attachInterrupt(digitalPinToInterrupt(P092_DLB_Pin), P092_Pin_changed, CHANGE)`)
>   by another interrupt?

The iRAM is quite small as in every now and then we run into issues where we cannot build our sketch.
I am not sure how big it actually is, but I do know there is not much left for our code to use.
It is also a bit confusing what iram actually is.
It is used to store executable code from flash, but some code (marked with the `ICACHE_RAM_ATTR`) and some variables (marked `volatile`) will remain pinned in that memory and never "swapped out" for other data or code.

Any way it is really important to have a good look at what to declare to remain pinned in there.


I don't know enough about priority on interrupts.
It is very well possible some interrupt does interfere with other interrupts.
I have seen some code specifically disabling serial interrupts for example, 
Maybe you have already seen this short intro on interrupts by random nerd tutorials: [ESP8266 Interrupts and Timers using Arduino IDE (NodeMCU)](https://randomnerdtutorials.com/interrupts-timers-esp8266-arduino-ide-nodemcu/)

You do call `yield()` in your loop.
The execution time of a yield is not constant and I can imagine it may interfere with the interrupt handling.
Maybe you can find some time in the decoding of the data where it is less of a problem if you miss a beat. For example between 2 bytes instead of 2 bits.

Reply:
Thanks for your answers.
I only call `yield()` to keep the ESP at life (watchdag and so on). The excution time does not matter because the bit change detection is handled by the interrupt function.
The waiting loop is only used to avoid any distortion of the bit change receiving.
Nothing else is done while looping and waiting.
The decoding of the received bit changes (stored in `volatile uint8_t DLbus_BitStream[DLbus_MaxDataBits]`)  is already outside of the interrupt and does not have a wait loop.

Reply:
> Nothing else is done while looping and waiting.

That's a bit of my point. It is not uncommon to have more sensors operational.

Reply:
I'll try another code as soon as possible.
By the way what is the best variable definition for `uint8_t DLbus_BitStream[DLbus_MaxDataBits]` used in the interrupt function and outside of the interrupt. I'll try to access this array outside of the interrupt only while the interrupt is paused.

Reply:
If you need it in an ISR function, then it must be declared volatile (or inside the `ICACHE_RAM_ATTR` function).
Not sure if declaring a pointer volatile and assigning a heap allocated vector to it is treated the same, although it should be.

Do you store bits in it, or do you compact it into bytes? (hopefully the last)
Please make sure not to allocate more than needed, so if the size of the array describes the number of compacted bits, then don't call it something referred to as bits.

Reply:
Oh and if you call some function from the ISR call, then that function also needs to have the `ICACHE_RAM_ATTR`

Reply:
I declared only variables needed in the interrupt function as `volatile`.
Do you have a code snippet for your proposal:
_declaring a pointer volatile and assigning a heap allocated vector to it_
No other function is called from the ISR call!
The received Bits are extendend to uint8_t, because they are extended by the timing flags needed for the analyze and processing functions.
I call it still BitStream but there is a comment to it.
`volatile uint8_t DLbus_BitStream[DLbus_MaxDataBits];  // received bit stream (each bit is extended to uint8_t, containing the timing flags)`, but it is also possible to use another name more related to bytes.

Reply:
as long as the data is not stored as 1 bit per byte, then it is OK :)
Like I said, I only did have a very quick look at the code and only commented on what caught my eye immediately.
As I said, the name of the allocation sounded like a bit per byte storage, but if you also encode timing in it, then it sounds like good use of the array.

My remark about the extra functions was with a simple idea I had in mind, to pack the data in a separate function, but then it would have been called from within the ISR function and...
Well as always, my mind is running way faster than I can type, which makes things often a bit more confusing for the person reading my replies. (sorry 'bout that)